File: Daily_challenge/image-overlap.cpp
#include<iostream>
#include<vector>
#include<unordered_map>
using namespace std;

class Solution{
    public:
    int largestOverlap(vector<vector<int>>& A, vector<vector<int>>& B) {
        int len = A.size();
        vector<pair<int, int>> a, b;
        for(int i = 0; i < len; i++){
            for(int j = 0; j < len; j++){
                if(A[i][j]) a.push_back({i, j});
                if(B[i][j]) b.push_back({i, j});
            }
        }
        unordered_map<int, int> mp;
        for(auto& x: a){
            for(auto& y: b){
                mp[(x.first - y.first) * 100 + (x.second - y.second)]++;
            }
        }
        int res = 0;
        for(auto& x: mp) res = max(res, x.second);
        return res;
    }
};

int main(){
    vector<vector<int>> A = {{1, 1, 0}, {0, 1, 0}, {0, 1, 0}};
    vector<vector<int>> B = {{0, 0, 0}, {0, 1, 1}, {0, 0, 1}};
    Solution s;
    cout << s.largestOverlap(A, B) << endl;
    return 0;
}

File: Daily_challenge/image-overlap.py
import numpy as np

class List:
    def __init__(self, x, y):
        self.x = x
        self.y = y

class Solution:
    def largestOverlap(self, A: List[List[int]], B: List[List[int]]) -> int:
        A = np.array(A)
        B = np.array(B)
        n = len(A)
        max_overlap = 0
        for i in range(n):
            for j in range(n):
                max_overlap = max(max_overlap, np.sum(A[i:, j:] * B[:n-i, :n-j]))
                max_overlap = max(max_overlap, np.sum(A[:n-i, j:] * B[i:, :n-j]))
                max_overlap = max(max_overlap, np.sum(A[i:, :n-j] * B[:n-i, j:]))
                max_overlap = max(max_overlap, np.sum(A[:n-i, :n-j] * B[i:, j:]))
        return max_overlap

if __name__ == '__main__':
    A = [[1,1,0],[0,1,0],[0,1,0]]
    B = [[0,0,0],[0,1,1],[0,0,1]]
    solution = Solution()
    print(solution.largestOverlap(A, B))